### PR TITLE
[1045] Providers should return List<?> instead of List<Object> whenever possible

### PR DESCRIPTION
**Pull request by ldelaigue**
Miscellaneous improvements to make it easier for applications to use sirius-components.

### CODE REVIEW

File: backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/EMFValidationService.java [hunk 1]
```diff
@@ -39,6 +40,7 @@
  * @author gcoutable
  */
 @Service
+@ConditionalOnMissingBean
```
Comment (sbegaudeau) [inline backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/EMFValidationService.java:43]: Why do you need that to be conditional? If you are using EMF, why wouldn't you use the EMF validation service?

File: backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/ObjectService.java [hunk 2]
```diff
@@ -47,6 +48,7 @@
  * @author hmarchadour
  */
 @Service
+@ConditionalOnMissingBean
```
Comment (sbegaudeau) [inline backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/ObjectService.java:51]: Same question here, if you need the EMF part of Sirius Components, why not use its object service?

**Review by sbegaudeau — CHANGES_REQUESTED** [review]

File: backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/EMFValidationService.java (hunk 1 above)
Comment (ldelaigue) [reply backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/EMFValidationService.java:43]: I'm not using EMF. The addition of these annotations make it possible for my application not to depend on EMF, that's the point :-)

File: backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/ObjectService.java (hunk 2 above)
Comment (ldelaigue) [reply backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/ObjectService.java:51]: I think I begin to see what you mean... I'm trying to not depend on EMF. How can I avoid having the services & other Spring components of sirius-components-emf scanned? Marking them as @ConditionalOnMissingBean works but may not be the best answer.
However, I still believe that even if I want to depend on EMF, it's be legitimate for an app to be able to inject their own implementation (which can be achieved by annotating it with @Primary)

File: backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/description/TreeDescription.java [hunk 3]
```diff
@@ -51,9 +51,9 @@
 
     private Function<VariableManager, Boolean> deletableProvider;
 
-    private Function<VariableManager, List<Object>> elementsProvider;
+    private Function<VariableManager, List<?>> elementsProvider;
```
[the remark this answers is not in the document]
Comment (ldelaigue) [reply backend/sirius-components-trees/src/main/java/org/eclipse/sirius/components/trees/description/TreeDescription.java:54]: I've created bug #1045; I will update the changelog as well

File: doc/adrs/041_use_wildcard_collections.adoc [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+** `SelectDescription`
+*** `optionsProvider`: Only used in `SelectComponent`, to transform into a list
+
+=== Compatibility layer
+
+N/A
+
+=== Frontend
+
+N/A
```
Comment (sbegaudeau) [inline doc/adrs/041_use_wildcard_collections.adoc:49]: Those parts are not required in every ADR, you can delete them

File: backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/EMFValidationService.java (hunk 1 above)
Comment (ldelaigue) [reply backend/sirius-components-emf/src/main/java/org/eclipse/sirius/components/emf/services/EMFValidationService.java]: I removed these useless @ConditionalOnMissingBean.

File: doc/adrs/041_use_wildcard_collections.adoc (hunk 4 above)
Comment (ldelaigue) [reply doc/adrs/041_use_wildcard_collections.adoc:49]: Okay

File: doc/adrs/041_use_wildcard_collections.adoc [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+** `ListDescription`
+*** `itemsProvider`: Only used in `ListComponent`, to transform into a list
+** `MultiSelectDescription`
+*** `optionsProvider`: Only used in `MultiSelectComponent`, to transform into a list
+** `PageDescription`
+*** `semanticElementsProvider`: Only used in `PageComponent`, to transform into a list
+** `RadioDescription`
+*** `optionsProvider`: Only used in `RadioComponent`, to transform into a list
+** `SelectDescription`
+*** `optionsProvider`: Only used in `SelectComponent`, to transform into a list
```
[the remark this answers is not in the document]
Comment (ldelaigue) [reply doc/adrs/041_use_wildcard_collections.adoc:41]: I'll update this with the exact list of the following commit

[the remark this answers is not in the document]
Comment (ldelaigue) [reply doc/adrs/041_use_wildcard_collections.adoc:41]: I'll also take into account `SelectionDescription.objectsProvider` & `ValidationDescription.diagnosticsProvider`

File: CHANGELOG.adoc [hunk 6]
```diff
@@ -6,6 +6,7 @@
 
 - [ADR-039] Provide a variable to detect the environment
 - [ADR-040] Add support for a post selection
+- [ADR-041] Use wildcard collections instead of List<Object> in Providers
```
[the remark this answers is not in the document]
Comment (ldelaigue) [reply CHANGELOG.adoc]: Done. I've updated this PR and pushed PRs to depending projects, including sirius-web.

File: doc/adrs/042_use_wildcard_collections.adoc [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+* Using `Iterable<?>`
+** Clarifies that what's returned is only intended to be iterated over, but does not make it possible to easily know the number of elements, or if there's an element at all.
+* Using `Collection<?>`
+** Better than `List<Object>`, does not impose ordering of elements, and makes it easy to check isEmpty() or size().
+* Using `List<?>`
+** Better than `List<Object>`, conveys that ordering of elements matters, and makes it easy to check isEmpty() or size().
+
+== Decision
+
```
Comment (pcdavid) [inline doc/adrs/042_use_wildcard_collections.adoc:20]: You mention several alternatives above, but the actrual "Decision" section does not make it clear which one was chosen.
Yes it's more or less implied by the properties of the alternatives, but it does not cost much to make the choice more explicit.

**Comment by pcdavid** [comment]
This changes APIs, so it will require a version bump (and the corresponding switch on downstream projects like Sirius Web).

**Review by sbegaudeau — APPROVED** [review]

**Pull request merged**